test/iso/component-param.js

```diff
@@ -0,0 +1,179 @@
+    `),
+    );
+    el.setParams({bookmark: {id: 2, info: `updated`}});
+    await nextAnimationFrame();
+    expect(el.innerHTML).to.equal(
+      compactHtml(`
+      <div>{"id":2,"info":"updated"}</div>
+    `),
+    );
+  });
+
+  it(`respects defaultParams`, async () => {
+    const el = new ParamParentApp();
+    el.connectedCallback();
+    document.body.appendChild(el);
+    await nextAnimationFrame();
+    expect(el.childNodes[0].params.defaultString).to.equal(`defaultString`);
+    expect(el.childNodes[0].params.noDefaultString).to.equal(undefined);
+  });
+
+  it(`respects required field`, () => {
+    const el = new ParamParentApp();
+    el.setConfig(`updateSync`, true);
+    el.connectedCallback();
+    document.body.appendChild(el);
+    expect(() => el.update({requiredString: undefined})).to.throw(
+      Error,
+      `param 'requiredString' in ParamChild is required, undefined passed in`,
+    );
+  });
+
+  it(`respects required field`, () => {
+    expect(() => new DefaultAndRequiredParam()).to.throw(
+      Error,
+      `param 'requiredString' in DefaultAndRequiredParam cannot have both required and default`,
+    );
+  });
+
+  it(`throws error if non primitive used in param schema(class)`, () => {
+    expect(() => new NonPrimitiveTypeParamClass()).to.throw(
+      Error,
+      `Invalid type: NonPrimitiveTypeParamClass for param: A in paramSchema. Only ('Array' | 'String' | 'Boolean' | 'Number' | 'Object' | 'Function' | 'Map' | 'Set') is valid.`,
+    );
+  });
+
+  it(`throws error if non primitive used in param schema(string)`, () => {
+    expect(() => new NonPrimitiveTypeParamString()).to.throw(
+      Error,
+      `Invalid type: json for param: A in paramSchema. Only ('Array' | 'String' | 'Boolean' | 'Number' | 'Object' | 'Function' | 'Map' | 'Set') is valid.`,
+    );
+  });
+
+  it(`throws error if unkown param pass in child`, () => {
+    const el = new ExtraParamPassInChild();
+    el.setConfig(`updateSync`, true);
+    expect(() => el.connectedCallback()).to.throw(Error, `extra param 'extra' on ParamChild is not defined in schema`);
+  });
+
+  it(`hooks will be run`, async () => {
+    const el = new ParamParentApp();
+    el.connectedCallback();
+    document.body.appendChild(el);
+    const child = el.childNodes[0];
+    await nextAnimationFrame();
+    child._config.hooks = {preUpdate: sinon.stub()};
+    el.update({str: `abc`, num: 5, bool: true});
+    await nextAnimationFrame();
+    await nextAnimationFrame();
+    expect(child.getConfig(`hooks`).preUpdate.callCount).to.equal(1);
+    // no state for the child
+    expect(child.getConfig(`hooks`).preUpdate.firstCall.args[0]).to.equal(null);
+    // preUpdate hook
+    expect(child.getConfig(`hooks`).preUpdate.firstCall.args[1].str).to.equal(`abc`);
+    expect(child.getConfig(`hooks`).preUpdate.firstCall.args[1].num).to.equal(5);
+    expect(child.getConfig(`hooks`).preUpdate.firstCall.args[1].bool).to.equal(true);
+    sinon.restore();
```

Comment:
This won't get called if the test case throws an exception before this line. That's why you generally use `beforeEach`/`afterEach` for this kind of cleanup